### PR TITLE
Auth/[PM-1260] - Existing User - Login with Trusted Device (Flow 2)

### PR DESCRIPTION
**Pull request by JaredSnider-Bitwarden**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

[PM-1260](https://bitwarden.atlassian.net/browse/PM-1260)
- Update existing post SSO login logic to set the user's symmetric key via Trusted Device Decryption if a device key is present (i.e., a user has previously trusted the device and decrypted via master password entry)
- Routing logic updates to be done later as part of finishing up [PM-2759](https://bitwarden.atlassian.net/browse/PM-2759)

## Code changes

**`DeviceTrustCryptoService` Changes:**
- **libs/common/src/auth/services/device-trust-crypto.service.implementation.ts:** 
    - Finish writing `decryptUserKeyWithDeviceKey(...)`
        - Adds new logic which removes device keys if the decryption fails for any reason 
    - [PM-1378](https://bitwarden.atlassian.net/browse/PM-1378) - Fix `trustDevice()` logic to properly use the full `UserKey` during the creation of the `DevicePublicKey(UserKey)` so that the decrypted `UserKey` is the full 64 byte key which is required to decrypt the `UserKey(UserPrivateKey)` later on. 
- **libs/common/src/auth/services/device-trust-crypto.service.spec.ts:** 
    - Added tests for `decryptUserKeyWithDeviceKey(...)`

**Login Strategy Changes:**
- **libs/common/src/auth/login-strategies/login.strategy.ts:** 
    - Must persist user device key out of local storage into account on login for trusted device decryption to function (i.e., no device key - not possible to decrypt user key)
- **libs/common/src/auth/login-strategies/sso-login.strategy.ts:** 
    - Refactor existing `setUserKey(...)` logic into new `trySetUserKeyForKeyConnector()` method
    - Add new `trySetUserKeyWithDeviceKey(...)` method which will attempt to use new `DeviceTrustCryptoService.decryptUserKeyWithDeviceKey(...)` to get a decrypted user key in memory which will be used later on to decrypt the users vault
- **libs/common/src/auth/login-strategies/login.strategy.spec.ts:**
    - Update account setup test to consider newly added `keys` and `decryptionOptions` props
    - Test device key persistence 
- **libs/common/src/auth/login-strategies/sso-login.strategy.spec.ts:**
    - Test all relevant scenarios around `setUserKey(...)` changes / additions

**Platform Changes:**
- **libs/common/src/platform/services/state.service.ts:**
    - [PM-1378](https://bitwarden.atlassian.net/browse/PM-1378) - `getDeviceKey()` was not returning a properly instantiated `DeviceKey` after pulling it out of local storage so had to fix that by manually instantiating it
- **libs/common/src/platform/models/domain/account.ts:**
    - Add device key to `AccountKeys.fromJSON()` static method
- **libs/common/src/platform/models/domain/account-keys.spec.ts:**
    - Add tests to `AccountKeys()` class tests for `deviceKey` persistence and JSON deserialization

**Auth Service Dependency Updates:**
- **libs/common/src/auth/services/auth.service.ts:**
    - Added `DeviceTrustCryptoService` to constructor dependencies
    - Browser
        - Required creating `browser/src/auth/background/service-factories/device-trust-crypto-service.factory.ts`
        - Required creating `browser/src/background/service-factories/devices-api-service.factory.ts`
        - Required updating `browser/src/auth/background/service-factories/auth-service.factory.ts` with new dependency factory 
        - Required updating the manual creation of the `AuthService` in `main.background.ts` 
    - CLI
        - Required updating `AuthService` dependencies in `bw.ts`

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

[PM-1260]: https://bitwarden.atlassian.net/browse/PM-1260?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PM-2759]: https://bitwarden.atlassian.net/browse/PM-2759?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[PM-1378]: https://bitwarden.atlassian.net/browse/PM-1378?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/common/src/auth/login-strategies/login.strategy.spec.ts [hunk 1]
```diff
@@ -174,11 +196,36 @@ describe("LogInStrategy", () => {
...
             },
           },
+          keys: new AccountKeys(),
+          decryptionOptions: acctDecryptionOptions,
         })
       );
       expect(messagingService.send).toHaveBeenCalledWith("loggedIn");
     });
 
+    it("for trusted device encryption, it persists a device key if it exists on login", async () => {
```
Comment (coroiu) [inline libs/common/src/auth/login-strategies/login.strategy.spec.ts:206]: nit: 
```suggestion
    it("persists a device key for trusted device encryption when it exists on login", async () => {
```

**Comment by coroiu** [comment]
Found multiple bugs: 

**1. Client is not able to decrypt vault items after logout/login with trusted device**
image

Reproduction steps:

1. Login with an SSO user that has never logged in before (to cause auto enrollment)
2. Trust device and continue to vault
3. Create a new vault item with name and some notes (or something)
4. Log out
5. Log back in again 
6. Observe that vault item cannot be decrypted, console shows log in screenshot above

**2. Seems like organization keys are not properly set**
image


1. Login with an SSO user that has never logged in before (to cause auto enrollment)
2. Trust device and continue to vault
3. Create a new vault item with name and some notes (or something)
4. Log out
5. Log in with admin and confirm new user into organization
6. Log out
7. Log back in with the new SSO user
8. Observe that the user is not a part of the organization, but client cannot decrypt and complains that private key is missing

File: libs/common/src/auth/login-strategies/sso-login.strategy.ts [hunk 2]
```diff
@@ -80,39 +82,68 @@ export class SsoLogInStrategy extends LogInStrategy {
...
-        await this.cryptoService.setUserKey(userKey);
-      }
+  private async trySetUserKeyWithDeviceKey(tokenResponse: IdentityTokenResponse): Promise<void> {
+    const trustedDeviceOption = tokenResponse.userDecryptionOptions?.trustedDeviceOption;
+
+    const deviceKey = await this.deviceTrustCryptoService.getDeviceKey();
+    const encDevicePrivateKey = trustedDeviceOption?.encryptedPrivateKey;
+    const encUserKey = trustedDeviceOption?.encryptedUserKey;
+
+    if (!deviceKey || !encDevicePrivateKey || !encUserKey) {
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/auth/login-strategies/sso-login.strategy.ts:120]: TBH, I find the first version easier to read without missing the negation.

File: libs/common/src/auth/services/device-trust-crypto.service.implementation.ts [hunk 3]
```diff
@@ -97,28 +98,40 @@ export class DeviceTrustCryptoService implements DeviceTrustCryptoServiceAbstrac
...
+
+      // Attempt to decrypt encryptedUserDataKey with devicePrivateKey
+      const userKey = await this.cryptoService.rsaDecrypt(
+        encryptedUserKey.encryptedString,
+        devicePrivateKey
+      );
+      return new SymmetricCryptoKey(userKey) as UserKey;
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.warn("TDE: Failed to decrypt user key with device key. Removing device key.", e);
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/auth/services/device-trust-crypto.service.implementation.ts:129]: I wasn't sure if this was allowed, but Jake had mentioned that we had this type of logic in the crypto service elsewhere so this was fine. I can remove if there is a hard rule against any debug warnings for certain error states in prod.

File: libs/common/src/auth/models/response/user-decryption-options/user-decryption-options.response.ts [hunk 4]
```diff
@@ -1,14 +1,26 @@
...
+import {
+  ITrustedDeviceUserDecryptionOptionServerResponse,
+  TrustedDeviceUserDecryptionOptionResponse,
+} from "./trusted-device-user-decryption-option.response";
+
+export interface IUserDecryptionOptionsServerResponse {
+  HasMasterPassword: boolean;
+  TrustedDeviceOption?: ITrustedDeviceUserDecryptionOptionServerResponse;
+  KeyConnectorOption?: IKeyConnectorUserDecryptionOptionServerResponse;
+}
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/auth/models/response/user-decryption-options/user-decryption-options.response.ts:16]: It was just the shortest way I could indicate they were interfaces.

File: libs/common/src/platform/services/state.service.ts [hunk 5]
```diff
@@ -1302,7 +1302,14 @@ export class StateService<
 
     const account = await this.getAccount(options);
 
-    return account?.keys?.deviceKey as DeviceKey;
+    const existingDeviceKey = account?.keys?.deviceKey;
+
+    // Must manually instantiate the SymmetricCryptoKey class from the JSON object
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/platform/services/state.service.ts:1307]: It is not deserialized into an instance of the `SymmetricCryptoKey` class when it comes out of state. The top level `Account.fromJSON()` does not run which leaves the object as:

```TypeScript
{
    base64: "..."
}
```

Which cannot be used for decryption.

File: libs/common/src/platform/models/domain/account-keys.spec.ts [hunk 6]
```diff
@@ -4,6 +4,7 @@ import { Utils } from "../../misc/utils";
 import { AccountKeys, EncryptionPair } from "./account";
 import { SymmetricCryptoKey } from "./symmetric-crypto-key";
 
+// TODO: add device key to tests
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/platform/models/domain/account-keys.spec.ts:7]: Argh, yes! Thank you for catching that!

File: libs/common/src/auth/services/device-trust-crypto.service.implementation.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (coroiu) [reply libs/common/src/auth/services/device-trust-crypto.service.implementation.ts:129]: If we want the logging in prod why not use `LogService` ?

[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/auth/services/device-trust-crypto.service.implementation.ts:129]: That's a fair point. For now, I'm going to remove it.

**Comment by JaredSnider-Bitwarden** [comment]
> Found multiple bugs:
> 
> **1. Client is not able to decrypt vault items after logout/login with trusted device** image
> 
> Reproduction steps:
> 
> 1. Login with an SSO user that has never logged in before (to cause auto enrollment)
> 2. Trust device and continue to vault
> 3. Create a new vault item with name and some notes (or something)
> 4. Log out
> 5. Log back in again
> 6. Observe that vault item cannot be decrypted, console shows log in screenshot above


This has been resolved:

https://github.com/bitwarden/clients/assets/116684653/3693ea33-1228-40b6-8835-1dfbc5952ef4


> **2. Seems like organization keys are not properly set** image
> 
> 1. Login with an SSO user that has never logged in before (to cause auto enrollment)
> 2. Trust device and continue to vault
> 3. Create a new vault item with name and some notes (or something)
> 4. Log out
> 5. Log in with admin and confirm new user into organization
> 6. Log out
> 7. Log back in with the new SSO user
> 8. Observe that the user is not a part of the organization, but client cannot decrypt and complains that private key is missing

This has been partially resolved. 

https://github.com/bitwarden/clients/assets/116684653/9a010fe1-cc32-445a-9f0d-9951b21f7ce2

Users no longer get errors for the private key missing, but the user is not getting the organization ciphers for some reason.

File: libs/common/src/auth/models/response/user-decryption-options/user-decryption-options.response.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/auth/models/response/user-decryption-options/user-decryption-options.response.ts:16]: I left this open as I'm not sure if this is an issue or not.

**Review by coroiu — APPROVED** [review]
🔥

File: apps/browser/src/background/service-factories/devices-api-service.factory.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+export type DevicesApiServiceInitOptions = DevicesApiServiceFactoryOptions & ApiServiceInitOptions;
+
+export function devicesApiServiceFactory(
+  cache: { DevicesApiService?: DevicesApiServiceAbstraction } & CachedServices,
+  opts: DevicesApiServiceInitOptions
+): Promise<DevicesApiServiceAbstraction> {
+  return factory(
+    cache,
+    "DevicesApiService",
```
Comment (justindbaur) [inline apps/browser/src/background/service-factories/devices-api-service.factory.ts:24]:
```suggestion
  cache: { devicesApiService?: DevicesApiServiceAbstraction } & CachedServices,
  opts: DevicesApiServiceInitOptions
): Promise<DevicesApiServiceAbstraction> {
  return factory(
    cache,
    "devicesApiService",
```

**Review by justindbaur — CHANGES_REQUESTED** [review]
Very good stuff, one tiny request.

**Review by justindbaur — APPROVED** [review]

**Pull request merged**